appengine/genetics/js/cage.js

```diff
@@ -188,55 +215,92 @@ Genetics.Cage.reset = function() {
- * @param {Function} doneCallback Function to call when game ends.
- */
-Genetics.Cage.start = function(doneCallback) {
-  Genetics.Cage.doneCallback_ = doneCallback;
-  // Create mice for each player.
-  for (var playerId in Genetics.Cage.players) {
-    for (var i = 0; i < Genetics.Cage.START_MICE_PER_PLAYER; i++) {
-      var mouseSex = i % 2 == 0 ? Genetics.Mouse.Sex.MALE :
-          Genetics.Mouse.Sex.FEMALE;
-      var mouseId = Genetics.Cage.nextAvailableMouseId_++;
-      var mouse = new Genetics.Mouse(mouseId, mouseSex, null, null, playerId);
-      Genetics.Cage.born(mouse);
-      new Genetics.Cage.Event('ADD', goog.object.clone(mouse),
-          Genetics.Cage.players[playerId][0]).addToQueue();
+ * @param {!Function} checkForEnd Function to call when game ends.
+ * that should be skipped in simulateLife.
+ */
+Genetics.Cage.start = function(checkForEnd) {
+  Genetics.Cage.checkForEnd = checkForEnd;
+  Genetics.Cage.stopped_ = false;
+  Genetics.Cage.nextAvailableMouseId_ = Genetics.Cage.nextRoundMice_.length;
+  for (var playerId = 0, player; player = Genetics.Cage.players[playerId];
+      playerId++) {
+    if (goog.isFunction(player.code)) {
+      // Cache code if player code is generator.
+      player.cachedCode = Genetics.Cage.players[1].code + '\n' + player.code();
     }
   }
-  Genetics.Cage.endTime_ = Date.now() + Genetics.Cage.GAME_TIME_LIMIT_MSEC;
   new Genetics.Cage.Event('START_GAME').addToQueue();
+
+  Genetics.Cage.update();
 };
 
 /**
- * Queues life simulation for mouse.
- * @param {!Genetics.Mouse} mouse The mouse to queue the simulation for.
+ * Whether the player mouse (mouse id 0) will be the only mouse running logic.
+ * @type {boolean}
+ */
+Genetics.Cage.soloPlayerMouse = false;
+
+/**
+ * Whether the player mouse (mouse id 0) will be the only mouse not running
+ * logic.
+ * @type {boolean}
  */
-Genetics.Cage.life = function(mouse) {
-  // Queue life simulation for mouse with some variation in timeout to allow for
-  // variation in order.
-  Genetics.Cage.lifePidsMap_[mouse.id] =
-      setTimeout(goog.partial(Genetics.Cage.simulateLife, mouse),
-          10 * Math.random());
+Genetics.Cage.ignorePlayerMouse = false;
+
+/**
+ * Whether pickFight will be skipped.
+ * @type {boolean}
+ */
+Genetics.Cage.skipPickFight = false;
+
+/**
+ * Executes the next mouse's action and checks for end game.
+ */
+Genetics.Cage.update = function() {
+  if (Genetics.Cage.HEADLESS || Genetics.Cage.keepHistory ||
+      Genetics.Cage.Events.length < 100) {
+    // If the events queue is not backed up, then continue computing.
+    if (Genetics.Cage.currentRoundmice_.length == 0) {
+      if (Genetics.Cage.roundNumber < Genetics.Cage.MAX_ROUNDS) {
+        // Cycle to next round.
+        Genetics.Cage.currentRoundmice_ = Genetics.Cage.nextRoundMice_.slice(0);
+        new Genetics.Cage.Event('NEXT_ROUND').addToQueue();
+      }
+      Genetics.Cage.roundNumber++;
+    }
+    if (Genetics.Cage.currentRoundmice_.length) {
```

Comment:
Optional: I'd add a line break above this line, so that it isn't thought of as just an "else" clause to the previous one (which I'd initially assumed it to be), but is considered a new phase altogether. A comment might be appropriate.